src/ai/backend/manager/idle.py

```diff
@@ -307,24 +305,148 @@ class UtilizationIdleChecker(BaseIdleChecker):
+            await self._disable_timeout(session_id)
+        elif status == AppStreamingStatus.NO_ACTIVE_CONNECTIONS:
+            await self._update_timeout(session_id)
+
+    async def _disable_timeout(self, session_id: SessionId) -> None:
+        log.debug(f"ExecutionIdleChecker._disable_timeout({session_id})")
+        await self._redis.set(
+            f"session.{session_id}.last_execution", "0", exist=self._redis.SET_IF_EXIST
+        )
+
+    async def _update_timeout(self, session_id: SessionId) -> None:
+        log.debug(f"ExecutionIdleChecker._update_timeout({session_id})")
+        t = await self._redis.time()
+        await self._redis.set(
+            f"session.{session_id}.last_execution",
+            f"{t:.06f}",
+            expire=float(sum(self.cpu_util_series) / len(self.cpu_util_series))
+        )
+
+    async def _session_started_cb(
+        self,
+        context: None,
+        source: AgentId,
+        event: SessionStartedEvent,
+    ) -> None:
+        await self._update_timeout(event.session_id)
+
+    async def _execution_started_cb(
+        self,
+        context: None,
+        source: AgentId,
+        event: ExecutionStartedEvent,
+    ) -> None:
+        await self._disable_timeout(event.session_id)
+
+    async def _execution_exited_cb(
+        self,
+        context: None,
+        source: AgentId,
+        event: ExecutionFinishedEvent | ExecutionTimeoutEvent | ExecutionCancelledEvent,
+    ) -> None:
+        await self._update_timeout(event.session_id)
+
+    async def _do_idle_check(
+        self,
+        context: None,
+        source: AgentId,
+        event: DoIdleCheckEvent,
+    ) -> None:
+
+        try:
+            return await super()._do_idle_check(context, source, event)
+        finally:
+            pass
+
+    async def check_session(self, session: Row, dbconn: SAConnection) -> bool:
+        session_id = session["id"]
+        active_streams = await self._redis.zcount(
+            f"session.{session_id}.active_app_connections"
+        )
+        if active_streams is not None and active_streams > 0:
+            return True
+
+        redis = await aioredis.create_redis("redis://127.0.0.1:8111/")
+        redis_out = await redis.get(str(session["session_id"]), encoding=None)
+        try:
+            live_stat = msgpack.unpackb(redis_out)
+        except Exception:
+            return True
+
+        cpu_util_pct = float(live_stat["cpu_util"]["pct"])
+        mem_util_pct = float(live_stat["mem"]["pct"])
+        cuda_mem_util_pct = float(live_stat["cuda_mem"]["pct"])
+
+        interval = self.timer.interval
+
+        if "m" in self.window:
+            window = int(self.window.replace("m", ""))
+            window_size = (window * 60) / interval
+        elif "h" in self.window:
+            window = int(self.window.replace("h", ""))
+            window_size = (window * 60 * 60) / interval
+        else:
+            raise Exception("Undefined unit of window interval")
```

Comment:
Please refer to the utilization function to convert humanized string to `datetime` object.
https://github.com/lablup/backend.ai-common/blob/main/src/ai/backend/common/utils.py#L144